### PR TITLE
Refactor EndpointInfo in Swift

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR converts the Swift EndpointInfo into classes like in C++ and C#.

### CODE REVIEW

File: swift/src/Ice/CommunicatorI.swift [hunk 1]
```diff
@@ -15,10 +15,12 @@ class CommunicatorI: LocalObject<ICECommunicator>, Communicator {
         self.initData = initData
         do {
             classGraphDepthMax = try initData.properties!.getIcePropertyAsInt("Ice.ClassGraphDepthMax")
-            precondition(classGraphDepthMax >= 1 && classGraphDepthMax <= 0x7FFF_FFFF, "Ice.ClassGraphDepthMax must be >= 0 and <= 0x7FFF_FFFF")
+            precondition(
```
Comment (bernardnormier) [inline swift/src/Ice/CommunicatorI.swift:18]: We need to Swift format more often.

File: swift/src/Ice/Endpoint.swift [hunk 2]
```diff
@@ -2,77 +2,155 @@
 
 import Foundation
 
+/// The user-level interface to an endpoint.
+public protocol Endpoint: AnyObject, CustomStringConvertible {
```
Comment (bernardnormier) [inline swift/src/Ice/Endpoint.swift:6]: Moved at the top of Endpoint.swift. Didn't change this protocol.

File: swift/src/Ice/Endpoint.swift [hunk 3]
```diff
@@ -2,77 +2,155 @@
...
+        underlying?.secure() ?? false
+    }
 
-/// The user-level interface to an endpoint.
-public protocol Endpoint: AnyObject, CustomStringConvertible {
-    /// Return a string representation of the endpoint.
-    ///
-    /// - returns: `String` - The string representation of the endpoint.
-    func toString() -> String
+    public init(underlying: EndpointInfo) {
```
Comment (bernardnormier) [inline swift/src/Ice/Endpoint.swift:50]: Logically "protected", but this is not available in Swift.

File: swift/src/IceImpl/Endpoint.mm [hunk 4]
```diff
@@ -64,82 +39,70 @@ + (id)createEndpointInfo:(std::shared_ptr<Ice::EndpointInfo>)infoPtr
...
-    if (std::dynamic_pointer_cast<Ice::SSL::EndpointInfo>(infoPtr))
-    {
-        return [factory createSSLEndpointInfo:handle
-                                   underlying:underlying
-                                      timeout:infoPtr->timeout
-                                     compress:infoPtr->compress];
-    }
-
-#if TARGET_OS_IPHONE
+#if TARGET_OS_IPHONE // TODO: remove conditional
```
Comment (bernardnormier) [inline swift/src/IceImpl/Endpoint.mm:76]: The IceIAP::EndpointInfo and ConnectionInfo classes are tiny classes that don't need be conditional.

File: swift/src/IceImpl/Endpoint.mm [hunk 5]
```diff
@@ -64,82 +39,70 @@ + (id)createEndpointInfo:(std::shared_ptr<Ice::EndpointInfo>)infoPtr
...
+        NSData* rawBytes = [[NSData alloc] initWithBytes:opaqueInfo->rawBytes.data()
+                                                  length:opaqueInfo->rawBytes.size()];
+
+        return [factory createOpaqueEndpointInfo:opaqueInfo->type()
+                                   encodingMajor:opaqueInfo->rawEncoding.major
+                                   encodingMinor:opaqueInfo->rawEncoding.minor
+                                        rawBytes:rawBytes];
+    }
+
+    // TODO: we should never return null from this factory function.
```
Comment (bernardnormier) [inline swift/src/IceImpl/Endpoint.mm:105]: Returning null will result in a crash.

File: swift/src/IceImpl/include/Endpoint.h [hunk 6]
```diff
@@ -9,74 +9,48 @@ ICEIMPL_API @interface ICEEndpoint : ICELocalObject
...
 #ifdef __cplusplus
 
 @interface ICEEndpoint ()
 @property(nonatomic, readonly) std::shared_ptr<Ice::Endpoint> endpoint;
-+ (nullable ICEEndpointInfo*)createEndpointInfo:(std::shared_ptr<Ice::EndpointInfo>)infoPtr NS_RETURNS_RETAINED;
-@end
-
-@interface ICEEndpointInfo ()
-@property(nonatomic, readonly) std::shared_ptr<Ice::EndpointInfo> info;
++ (id)createEndpointInfo:(std::shared_ptr<Ice::EndpointInfo>)infoPtr NS_RETURNS_RETAINED;
```
Comment (bernardnormier) [inline swift/src/IceImpl/include/Endpoint.h:53]: I am not sure about this signature, in particular the NS_RETURNS_RETAINED. Is this correct?

The object we return is a Swift object. It does not (but could) derive from NSObject.

**Review by pepone — APPROVED** [review]

**Pull request merged**